onlinejudge/service/leetcode.py

```diff
@@ -0,0 +1,264 @@
+            'operationName': 'globalData',
+            'query': '\n'.join([
+                'query globalData {',
+                '    userStatus {',
+                '        isSignedIn',
+                '    }',
+                '}',
+            ]),
+        }
+        resp = utils.request('POST', 'https://leetcode.com/graphql', session=session, json=json_body)
+        json_resp = resp.json()
+        return json_resp['data']['userStatus']['isSignedIn']
+
+
+class LeetCodeProblem(onlinejudge.type.Problem):
+    """
+    :ivar title_slug: :py:class:`str`
+    """
+    def __init__(self, *, title_slug: str):
+        self.title_slug = title_slug
+
+    def _set_request_header(self, session: Optional[requests.Session] = None) -> requests.Session:
+        service = self.get_service()
+        session = service._set_request_header(session)
+        service_url = service.get_url()
+        session.headers.update({
+            'Referer': f'{service_url}problems/{self.title_slug}/',
+        })
+        return session
+
+    # TODO: enable to get premium only questions as well
+    def download_sample_cases(self, *, session: Optional[requests.Session] = None) -> List[TestCase]:
+        session = self._set_request_header(session)
+        json_body = {
+            'operationName': 'getQuestionDetail',
+            'query': '\n'.join([
+                'query getQuestionDetail($titleSlug: String!) {',
+                '    question(titleSlug: $titleSlug) {',
+                '        content',
+                '    }',
+                '}',
+            ]),
+            'variables': {
+                'titleSlug': self.title_slug
+            },
+        }
+
+        resp = utils.request('POST', 'https://leetcode.com/graphql', session=session, json=json_body)
+        json_resp = resp.json()
+        content_html = json_resp['data']['question']['content']
+        if content_html is None:
+            logger.warning("This problem seems to be locked: need premium?")
+            return []
+        soup = bs4.BeautifulSoup(content_html, utils.html_parser)
+        test_cases = []
+
+        for num, pre in enumerate(soup.find_all('pre')):
+            children = pre.contents
+            idx, input_data, output_data = 0, '', ''
+
+            # find input data
+            while (idx < len(children) and (children[idx].name != 'strong' or len(children[idx].contents) != 1 or 'input' not in children[idx].contents[0].lower())):
+                idx += 1
+            idx += 1
+            if idx < len(children):
+                input_data = children[idx].strip()
+
+            # find output data
+            while (idx < len(children) and (children[idx].name != 'strong' or len(children[idx].contents) != 1 or 'output' not in children[idx].contents[0].lower())):
+                idx += 1
+            idx += 1
+            if idx < len(children):
+                output_data = children[idx].strip()
+
+            if input_data and output_data:
```

Comment:
なるほどです。
こちら如何せん人が読む問題文中から抽出しているため、LeetCode側の不備でブレやエラーがありそうだなと思い、このような実装にしていました。実際に第1問が

> Input: nums = [2,7,11,15], target = 9
> Output: [0,1]
> Output: Because nums[0] + nums[1] == 9, we return [0, 1].

というふうにOutputが2つ書かれていて、2つめはOutputではなく説明になっていたりします。
（第2問では

> Input: l1 = [2,4,3], l2 = [5,6,4]
> Output: [7,0,8]
> Explanation: 342 + 465 = 807.

のようになっています）

ほとんどの場合、テストケースは

> Input: ...
> Output: ...
> Explanation: ...

もしくは

> Input: ...
> Output: ...

と書かれているかと思うので、これを想定して、そうでない場合にErrorとするのもありかなと思うのですが、その場合、いきなり第1問のサンプルケースが取得できないこともあり、判断に迷っています。

Reply:
第1問からエラーになるのはつらいというのはそうですね。どういうフォーマットがどれくらいの割合なのかは私は分からない (LeetCode にはほとんど参加していない) ので、警告の表示だけかエラーで落とすかの判断は @usk83 さんに任せます。